### PR TITLE
[#301] Python SyntaxErrors as CompileErrors

### LINKED ISSUE #301: Raise Python SyntaxErrors as CompileErrors
Currently all problems running Python code get treated as runtime errors. We should at least give a proper `CompilerError` if the ast parser chokes on any input.

### PR DESCRIPTION
**Pull request by indig**
Grabs `SyntaxError`s generated when parsing/compiling the `ast` and lifts them into `CompileErrors`. 

Resolves #301

### CODE REVIEW

File: polynote-kernel/src/main/scala/polynote/kernel/lang/python/PythonInterpreter.scala [hunk 1]
```diff
@@ -198,28 +229,32 @@ class PythonInterpreter(val kernelContext: KernelContext) extends LanguageInterp
...
-          jep.eval("__polynote_parsed__ = LastExprAssigner().visit(__polynote_parsed__)")
-          jep.eval("__polynote_parsed__ = ast.fix_missing_locations(__polynote_parsed__)")
-          jep.eval("__polynote_code__ = compile(__polynote_parsed__, '<ast>', 'exec')")
 
+        val pyResults = for {
+          _ <- wrapEval("__polynote_parsed__ = ast.parse(__polynote_code__, __polynote_cell__, 'exec')").right
+          _ <- wrapEval("__polynote_parsed__ = LastExprAssigner().visit(__polynote_parsed__)").right
+          _ <- wrapEval("__polynote_parsed__ = ast.fix_missing_locations(__polynote_parsed__)").right
+          _ <- wrapEval("__polynote_code__ = compile(__polynote_parsed__, '<ast>', 'exec')").right
+        } yield {
```
Comment (indig) [inline polynote-kernel/src/main/scala/polynote/kernel/lang/python/PythonInterpreter.scala]: There's probably a nicer way to handle this other than an `Either` whose `Right` we just throw away.

File: polynote-kernel/src/test/scala/polynote/kernel/lang/python/PythonInterpreterSpec.scala [hunk 2]
```diff
@@ -304,4 +304,12 @@ class PythonInterpreterSpec extends FlatSpec with Matchers with KernelSpec {
...
     }
   }
+
+  "foo" should "bar" in {
+    assertPythonOutput("this is a syntax error") { case (vars, output, displayed) =>
+      vars.toSeq shouldBe empty
+      output shouldBe empty
+      displayed should contain only "text/html" -> "hi"
+    }
+  }
```
Comment (indig) [inline polynote-kernel/src/test/scala/polynote/kernel/lang/python/PythonInterpreterSpec.scala]: Oh, this should be made into a full-blown test, I'll do that.

File: polynote-kernel/src/main/scala/polynote/kernel/lang/python/PythonInterpreter.scala [hunk 3]
```diff
@@ -198,28 +230,33 @@ class PythonInterpreter(val kernelContext: KernelContext) extends LanguageInterp
...
-        val numStats = jep.getValue("len(__polynote_parsed__.body)", classOf[java.lang.Long])
-
-        if (numStats > 0) {
-          jep.eval("__polynote_parsed__ = LastExprAssigner().visit(__polynote_parsed__)")
-          jep.eval("__polynote_parsed__ = ast.fix_missing_locations(__polynote_parsed__)")
-          jep.eval("__polynote_code__ = compile(__polynote_parsed__, '<ast>', 'exec')")
 
+        val pyResults = for {
+          _ <- wrapEval("__polynote_parsed__ = ast.parse(__polynote_code__, __polynote_cell__, 'exec')").right
+          _ <- (if (jep.getValue("len(__polynote_parsed__.body)", classOf[java.lang.Integer]) > 0) Right(List.empty[Result]) else Left(List.empty[CompileErrors])).right
```
Comment (indig) [inline polynote-kernel/src/main/scala/polynote/kernel/lang/python/PythonInterpreter.scala]: 😬

Comment (jeremyrsmith) [reply polynote-kernel/src/main/scala/polynote/kernel/lang/python/PythonInterpreter.scala]: If you import `cats.syntax.either._` then you don't need to say `.right` anymore

Comment (indig) [reply polynote-kernel/src/main/scala/polynote/kernel/lang/python/PythonInterpreter.scala]: Done, thanks for the tip!

**Pull request merged**